### PR TITLE
#57 adds marking for sewing

### PR DESCRIPTION
**Pull request by MikDal002**
I think it solves: #57 

UI:
image
Example output:
image

Works on folio
image

quatro
image

octavio
image

Link to excalidraw image: https://excalidraw.com/#json=0MM-vkb65hcBYfuZSDCYA,Z5gNvdXsQkJbhEmimFlcyg

### CODE REVIEW

**Comment by momijizukamori** [comment]
I probably won't be able to get to a review until later this week, but thank you for taking a stab at this!

**Comment by acestronautical** [comment]
I think "french link" need not be mentioned. Kettle and french link use the same hole layout AFAIK, so we can probably just generically say "stitching holes" or "sewing marks" or something of that nature.

File: src/book.js [hunk 1]
```diff
@@ -80,6 +89,12 @@ export class Book {
     this.page_positioning = configuration.pagePositioning;
     this.flyleafs = configuration.flyleafs;
     this.cropmarks = configuration.cropMarks;
+    this.frenchStichLink = {
```
Comment (acestronautical) [inline src/book.js:92]: Nit pick "Stich" is mispelled. I would prefer this be name stitchMarks or something similar. French Link is one specific stitch and not the only one.

File: src/utils/drawing.js [hunk 2]
```diff
@@ -112,6 +121,41 @@ export function drawCropmarks(papersize, per_sheet) {
...
+ */
+export function drawFrenchStichMarks(papersize, amount, marginPt, spacingPt) {
+  console.log("French marking!");
+  const [width, height] = papersize;
+  
+  const y = height * 0.5;
+  const commonCircleValues = {y: y, size: 1, color: grayscale(0.0)}
+
+  const workingWidth = width - 2 * marginPt;
+  const spaceBettwenPoints = workingWidth / (amount + 1);
```
Comment (acestronautical) [inline src/utils/drawing.js:139]: another nitpick, "Bettwen" is mispelled

Comment (MikDal002) [reply src/utils/drawing.js]: changed

**Comment by MikDal002** [comment]
> Collaborator

I've changed the naming and adjusted images in PR's description

**Comment by MikDal002** [comment]
> I think "french link" need not be mentioned. Kettle and french link use the same hole layout AFAIK, so we can probably just generically say "stitching holes" or "sewing marks" or something of that nature.
> 
> For the inputs we could maybe do "number of tapes" and "size of tapes" and assume that the spacing is always evenly distributed.

I've changed the naming and adjusted images in PR's description

**Comment by MikDal002** [comment]
> I probably won't be able to get to a review until later this week, but thank you for taking a stab at this!

Please consider, if functionality to print only on most inner page should be introduced? Because actually, points are printed on every page :).

**Comment by sithel** [comment]
🤔 know why the Preview isn't working? a compile error? https://github.com/momijizukamori/bookbinder-js/actions/runs/8693335860/job/23858049531?pr=113#step:5:16   (I've not yet checked out branch, just sat down at computer and have forgotten how to check out branches 🤦‍♀️ )

**Comment by sithel** [comment]
> Please consider, if functionality to print only on most inner page should be introduced? Because actually, points are printed on every page :).

🙌  I think if it was just inner, that'd be great! You should totally slap `isSigMiddle` in that object returned by `Signatures.booklet`!
(maybe inner & outer? Or let user select either/both? Don't let me feature creep your PR, this is great and am thrilled that you added it! I like the UX you've gone with)

**Comment by sithel** [comment]
Ah! could you bump the version number? https://github.com/momijizukamori/bookbinder-js/blob/main/package.json#L3

File: src/models/configuration.js [hunk 3]
```diff
@@ -94,6 +94,12 @@ export const schema = z.object({
   foreEdgePaddingPt: urlSafe(z.coerce.number()).default(0), // specific to wacky small
   wackySpacing, // specific to wacky small
   flyleafs: urlSafe(z.coerce.number()).default(1),
+
+  sewingMarksEnabled: urlSafe(coercedBoolean).default(false),
+  sewingMarksMarginPt: urlSafe(z.coerce.number()).default(72),
+  sewingMarksAmount: urlSafe(z.coerce.number()).default(3),
+  sewingMarksTapeWidthPt: urlSafe(z.coerce.number()).default(36)
```
Comment (momijizukamori) [inline src/models/configuration.js:101]: This line needs a comma at the end, that's why the build and preview are failing

Comment (MikDal002) [reply src/models/configuration.js:101]: mistake during merge conflicts :).

**Comment by MikDal002** [comment]
> > Please consider, if functionality to print only on most inner page should be introduced? Because actually, points are printed on every page :).
> 
> 🙌 I think if it was just inner, that'd be great! You should totally slap `isSigMiddle` in that object returned by `Signatures.booklet`! (maybe inner & outer? Or let user select either/both? Don't let me feature creep your PR, this is great and am thrilled that you added it! I like the UX you've gone with)

I actually understand what you meant by feature creep :D. As I started adding new settings, then everything got a little messy. What to make it clearer, I would need to introduce new layout for this feature. If you think it is okay, then  I can proceed :).

image

**Comment by MikDal002** [comment]
I have enabled actions on my fork to speed up resolution of problems. https://github.com/MikDal002/bookbinder-js/pull/1
Requiring Linux-like line endings makes problems on a Windows machine :D.

**Comment by momijizukamori** [comment]
> I have enabled actions on my fork to speed up resolution of problems. [MikDal002#1](https://github.com/MikDal002/bookbinder-js/pull/1) Requiring Linux-like line endings makes problems on a Windows machine :D.

I think you can configure git to auto-convert line endings - https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

Sorry about the mess with the preview action - I think it's security-related, but I didn't write the action, I'm just using it :c

File: src/utils/drawing.js [hunk 4]
```diff
@@ -112,6 +121,72 @@ export function drawCropmarks(papersize, per_sheet) {
...
+export function drawSewingMarks(sigDetails, position, amount, marginPt, tapeWidthPt) {
+  // Here normalize coordinates to always think in x an y like this
+  // | P        |H|    P |
+  // |  A       |E|   A  |
+  // |   G      |I|  G   |
+  // |    E     |G| E    |
+  // |          |T|      |
+  // |-POSITION-| |      |
+
+  // Left pages has spine position on the endge :/
```
Comment (sithel) [inline src/utils/drawing.js:141]: nice comment!

**Comment by MikDal002** [comment]
> > I have enabled actions on my fork to speed up resolution of problems. [MikDal002#1](https://github.com/MikDal002/bookbinder-js/pull/1) Requiring Linux-like line endings makes problems on a Windows machine :D.
> 
> I think you can configure git to auto-convert line endings - https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings
> 
> Sorry about the mess with the preview action - I think it's security-related, but I didn't write the action, I'm just using it :c

No problem, I've read the docs, and noticed that it is not working for forks :D. I am using autoconvert, but it is converting to `\r\n` on Windows machine where I start linter (which breaks because of line endings), then it gets converted just to `\n` when commit :). Co it keeps it consisted in repository :).

File: src/utils/drawing.js (hunk 4 above)
Comment (MikDal002) [reply src/utils/drawing.js:141]:
```suggestion
  // Left pages have spine position on the edge :/
```

Comment (MikDal002) [reply src/utils/drawing.js]: Changed spelling :)

Comment (sithel) [reply src/utils/drawing.js]: OH!!! omg - I meant I was genuinely excited by the documentation about page information!! 
absolutely did not notice spelling/grammar!! gah- I would never call out such things, even if I did notice them!
so sorry - really - just was trying to express happiness at code improving

Comment (MikDal002) [reply src/utils/drawing.js]: No problem :)

**Comment by MikDal002** [comment]
To make it clear, I've finished although you would like to see a proposal UI instead of that implemented :)

**Comment by sithel** [comment]
What you have is good! 
I think I feel good w/ the code, but want to run it/test it before I 👍 
unfortunately I can NOT figure out how to check out this PR (because forked) and the preview isn't working 😩 
it'll happen, this will work.. just... need a bit more time...

**Comment by MikDal002** [comment]
> What you have is good! I think I feel good w/ the code, but want to run it/test it before I 👍 unfortunately I can NOT figure out how to check out this PR (because forked) and the preview isn't working 😩 it'll happen, this will work.. just... need a bit more time...

Nice :) You can always check it on my branch here: https://mikdal002.github.io/bookbinder-js/pr-preview/pr-1/ and here https://github.com/MikDal002/bookbinder-js/pull/1. Notice SHAs of commits that they are the same here and in my PR :). Maybe it will help a little.

**Review by sithel — APPROVED** [review]
🙌  Thank you!

**Comment by sithel** [comment]
@MikDal002 - very excited to have this! 
Was there any follow-up/next work you had in mind? 

I think I'm going to try and build atop this in the next couple days to get my spine order PR in & have a pulled out "PDF markup section" -- heads up in case of conflicting plans?

**Comment by MikDal002** [comment]
> @MikDal002 - very excited to have this! 
> 
> Was there any follow-up/next work you had in mind? 
> 
> 
> 
> I think I'm going to try and build atop this in the next couple days to get my spine order PR in & have a pulled out "PDF markup section" -- heads up in case of conflicting plans?

Actually those changes are sufficient for me, already sawed two books thanks to your app and those changes, and next awaits :-). I could work on the ui part I've presented above, but from my understanding you are taking this idea, then great 😊. So I don't have anything to add from my side, although I can help in this project somehow?

**Pull request merged**